pkg/scheduler/core/equivalence/eqivalence.go

```diff
@@ -33,27 +33,20 @@ import (
 	"github.com/golang/glog"
 )
 
-// nodeMap stores a *Cache for each node.
-type nodeMap map[string]*NodeCache
-
 // Cache is a thread safe map saves and reuses the output of predicate functions,
 // it uses node name as key to access those cached results.
 //
 // Internally, results are keyed by predicate name, and "equivalence
 // class". (Equivalence class is defined in the `Class` type.) Saved results
 // will be reused until an appropriate invalidation function is called.
 type Cache struct {
-	// NOTE(harry): Theoretically sync.Map has better performance in machine with 8+ CPUs, while
-	// the reality is lock contention in first level cache is rare.
-	mu          sync.RWMutex
-	nodeToCache nodeMap
+	// i.e. map[string]*NodeCache
+	sync.Map
```

Comment:
If you look at the [sync.Map documentation](https://golang.org/pkg/sync/#Map), it reads:

"The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex."

eCache does not follow any of these patterns. The second pattern is closer to what eCache does, but even that is not completely applicable to eCache. We have informer event handlers which may write/delete eCache entries in parallel to other go-routines that run our predicate functions. So, the sets of entries written/deleted in various goroutines are not disjoint.
The documentation also states:
"The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content."

These are the reasons that I am a bit unsure about using sync.Map here. Also, the performance improvement does not seem to be large. Those said, I don't have any serious objection against it.

Reply:
Thanks Bobby! Actually, that's also the reason I did not use `sync.Map` at the beginning :D 

Since lock contention is visibility reduced during benchmark CPU profiling:  

https://github.com/resouer/temp/blob/master/torch_lock_true.5000.svg
https://github.com/resouer/temp/blob/master/torch_lock_true.map.5000.svg

and the Cache logic is quite simple, I guess we can keep this.